README.md
# connection_status_bar

A widget that animates when internet connection changes

![](demo.gif)

## Getting Started

add it to your dependencies then use it anywhere on your app, usually in a widget that is on top of all of your widgets.

```dart
ConnectionStatusBar(
     height : 25, // double: default height
     width : double.maxFinite, // double: default width
     color : Colors.redAccent, // Color: default background color
     lookUpAddress : 'google.com', // String: default site to look up for checking internet connection
     endOffset : const Offset(0.0, 0.0), // Offset: default animation finish point offset
     beginOffset : const Offset(0.0, -1.0), // Offset: default animation start point offset
     animationDuration : const Duration(milliseconds: 200), // Duration: default animation duration
     // Text: default text
     title : const Text(
       'Please check your internet connection',
       style: TextStyle(color: Colors.white, fontSize: 14),
     ),
),
```


CHANGELOG.md
## [1.1.6]

- Migrated from the deprecated connectivity plugin to connectivity_plus.

## [1.1.5] - 2021 14 April

- Null Safety

## [1.1.4] - 2020 2 January

- Add collapsedHeight

## [1.1.3] - 2020 2 January

- update connectivity dependency

## [1.1.2] - 2020 2 January

- update connectivity dependency

## [1.1.1] - 2019 12 October

- update look up docs

## [1.1.0] - 2019 12 October

- add ability to set look up address

## [1.0.0] - 2019 21 September

- first stable release
- add more customizable options

## [0.0.4] - 2019 15 September

- fix initial connection check

## [0.0.3] - 2019 15 September

- update docs

## [0.0.2] - 2019 15 September

- update docs
- change widget name

## [0.0.1] - 2019 15 September

- initial release


test/connection_status_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:connection_status_bar/connection_status_bar.dart';

void main() {
  test('todo add tests', () {});
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:connection_status_bar/connection_status_bar.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Material(
      child: Stack(
        children: [
          Align(
            alignment: Alignment.topCenter,
            child: Scaffold(
              appBar: AppBar(
                title: Text(widget.title),
              ),
              body: Center(
                child: Text(
                  'turn off device internet connection to see the widget animating',
                  textAlign: TextAlign.center,
                ),
              ),
            ),
          ),
          Align(
            alignment: Alignment.topCenter,
            child: ConnectionStatusBar(),
          ),
        ],
      ),
    );
  }
}


lib/connection_status_bar.dart
library connection_status_bar;

import 'dart:async';
import 'dart:io';

import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter/material.dart';

class ConnectionStatusBar extends StatefulWidget {
  final Color color;
  final Widget title;
  final double width;
  final double height;
  final double? collapsedHeight;
  final Offset endOffset;
  final Offset beginOffset;
  final Duration animationDuration;
  final String lookUpAddress;

  ConnectionStatusBar({
    Key? key,
    this.height = 25,
    this.collapsedHeight,
    this.width = double.maxFinite,
    this.color = Colors.redAccent,
    this.endOffset = const Offset(0.0, 0.0),
    this.beginOffset = const Offset(0.0, -1.0),
    this.animationDuration = const Duration(milliseconds: 200),
    this.lookUpAddress = 'google.com',
    this.title = const Text(
      'Please check your internet connection',
      style: TextStyle(color: Colors.white, fontSize: 14),
    ),
  }) : super(key: key);

  _ConnectionStatusBarState createState() => _ConnectionStatusBarState();
}

class _ConnectionStatusBarState extends State<ConnectionStatusBar> with SingleTickerProviderStateMixin {
  late StreamSubscription _connectionChangeStream;
  bool _hasConnection = true;
  late AnimationController controller;
  late Animation<Offset> offset;

  @override
  void initState() {
    _ConnectionStatusSingleton connectionStatus = _ConnectionStatusSingleton.getInstance();
    connectionStatus.initialize(widget.lookUpAddress);
    _connectionChangeStream = connectionStatus.connectionChange.listen(_connectionChanged);
    controller = AnimationController(vsync: this, duration: widget.animationDuration);

    offset = Tween<Offset>(begin: widget.beginOffset, end: widget.endOffset).animate(controller);
    super.initState();
  }

  void _connectionChanged(bool hasConnection) {
    if (_hasConnection == hasConnection) return;
    hasConnection == false ? controller.forward() : controller.reverse();
    _hasConnection = hasConnection;
  }

  @override
  Widget build(BuildContext context) {
    return SlideTransition(
      position: offset,
      child: Container(
        child: SafeArea(
          bottom: false,
          child: Container(
            color: widget.color,
            width: widget.width,
            height: _hasConnection ? widget.height : widget.collapsedHeight ?? widget.height,
            child: Center(
              child: widget.title,
            ),
          ),
        ),
      ),
    );
  }

  @override
  void dispose() {
    _connectionChangeStream.cancel();

    super.dispose();
  }
}

class _ConnectionStatusSingleton {
  String? _lookUpAddress;
  static final _ConnectionStatusSingleton _singleton = _ConnectionStatusSingleton._internal();
  _ConnectionStatusSingleton._internal();

  static _ConnectionStatusSingleton getInstance() => _singleton;

  bool hasConnection = true;

  StreamController<bool> connectionChangeController = StreamController.broadcast();

  final Connectivity _connectivity = Connectivity();

  void initialize(String lookUpAddress) {
    this._lookUpAddress = lookUpAddress;
    _connectivity.onConnectivityChanged.listen(_connectionChange);
    checkConnection();
  }

  Stream<bool> get connectionChange => connectionChangeController.stream;

  void dispose() {
    connectionChangeController.close();
  }

  void _connectionChange(ConnectivityResult result) {
    checkConnection();
  }

  Future<bool> checkConnection() async {
    assert(_lookUpAddress != null || _lookUpAddress != '');
    bool previousConnection = hasConnection;

    try {
      final result = await InternetAddress.lookup(_lookUpAddress!);
      if (result.isNotEmpty && result[0].rawAddress.isNotEmpty) {
        hasConnection = true;
      } else {
        hasConnection = false;
      }
    } on SocketException catch (_) {
      hasConnection = false;
    }

    if (previousConnection != hasConnection) {
      connectionChangeController.add(hasConnection);
    }

    return hasConnection;
  }
}
